Multiple TCP Connections for Migration

blitz opened the pull request:
# Summary

This PR implements VM migration using multiple TCP connections.

The send-migration HTTP command now accepts a `connections` parameter (defaults to 1) that specifies how many connections to use for migration. 

Ticket: https://github.com/cobaltcore-dev/cobaltcore/issues/205

# Design

If `connections` is larger than 1, the sender will attempt to establish additional TCP connections to the same migration destination. The main (initial) connection handles most of the migration protocol. Additional connections handle _only_ `Memory` commands for transferring chunks of VM memory.

We use a thread-per-connection model. This adds the burden of creating, keeping track of and tearing threads down. But given the I/O model of Cloud Hypervisor this seemed to be the only feasible option.

For each iteration of sending memory, the `MemoryTable` that describes dirty memory will be split into chunks of fixed size (`CHUNK_SIZE`). These chunks will then be distributed among the available threads in a round-robin fashion. Threads can have a configurable backlog of outstanding chunks to send (`BUFFERED_REQUESTS_PER_THREAD`). This is currently 1 to avoid accumulating too much backlog on connections that encountered throughput issues.

We still use the original request-response scheme. Since we don't pipeline requests, but always wait for the other side to acknowledge them, we have a fundamental limit on throughput that we can reach. The original code only expected one ACK for the whole dirty memory table. We now have one ACK per chunk.

I've come up with this formula of the upper bound of throughput per connection:

`effective_throughput = chunk_size / (chunk_size / throughput_per_connection + round_trip_time)`

This formula is also in the code. I've played around with this and with large enough chunks, the impact seems negligable, especially since we can scale up connections. Feel free to plug in your favorite numbers.

In a better world, we would pipeline requests and not wait for responses. The good thing: The sender is in full control over the parameters, so we can make them configurable if necessary. But I would rather keep this complexity from users.

# Noteworthy Issues

## New

- Teardown of threads will hang if the sender stops sending data on a connection. But then the whole VM migration will not succeed anyhow. This is difficult to fix without rewriting some of the I/O code and probably not worth it for now.

## Still There

- ~~The code scans the whole dirty bitmap with one thread with an unoptimized loop and then stores the result in a large vector. The dirty bitmap can reach hundreds of MB in size on huge instances. For lots of non-contiguous writes, the size of the resulting `MemoryTable` can be huge as well.~~ #36 
- The code only starts sending memory once the whole bitmap is scanned.

These issues give the VM time to mutate more memory and thus lengthen the time needed for migration.

# Noteworthy Changes in the Code

## More Consistent Error Handling

Previously, the protocol handling had some issues, such as not sending responses for unknown commands etc. Now we always send a response. Some errors resulted in error responses, some errors resulted in dying. This should all be consistent now and can be changed in one place.

What's still missing is to handle payloads for all migration commands. For now any payload bytes are ignored in the most inconvenient way (going out-of-sync between sender and receiver).

## State Machine for Receiver

Moved to #34.

You may ask where the additional connections come into play: When we enter the `Configured` state, the sender will establish more connections to send memory and tear them down, once we are done sending memory.

# Left to be done

- [x] Address first round of feedback
- [x] Fail gracefully if someone sends `connections: 0`
- [x] Split state machine into separate PR: #34
- [x] Merge #34 
- [x] Fail gracefully if the other end doesn't accept multiple connections
- [x] Update `src/api/openapi/cloud-hypervisor.yaml` for `connections` parameter 
- [ ] Add tests to the test suite
  - [ ] Test for multiple connections
  - [ ] Test for forward/backward compatibility? (maybe optional)
- [ ] Run test suite
  - [x] Regression tests (libvirt-tests, do we break usage that doesn't use multiple connections?)
  - [ ] With new tests
- [ ] Test on some big boxes with lots of guest RAM

blitz committed:
vm-migration: add connections field to HTTP API

This is not wired up to anywhere yet. We will use this to establish
multiple connections for live migration.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: inline functions in vm_receive_memory

This has no functional change, but it is a requirement to remove the
lock that used to obtain the MemoryManager instance.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: keep direct reference to guest memory around

... to avoid having to grab a lock when we receive a chunk of memory
over the migration socket. This will come in handy when we have
multiple threads for receiving memory.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: allow keeping the socket listener around

To allow for accepting more connections in the migration receive code
paths, we need to keep track of the listener. This commit adds a thin
abstraction to be able to hold on to it regardless of whether it is a
UNIX domain or TCP socket.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: handle receiving migration data on additional connections

We keep the listening socket around and accept as many connections as
the sender wants to open.

There are still some problems: We never tear these threads down
again. We will handle this in subsequent commits.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: use the same function to send memory

In anticipation of using multiple threads for sending memory, refactor
the sending code to be in a single place.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: move socket creation out of impl Vmm

... to be able to re-use it when establishing multiple send
connections. I moved the receive socket creation out for symmetry.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: move sending memory to the other migration code

... to simplify sending memory from multiple connections in future
commits.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: funnel memory data via additional connections abstraction

This does not actually use the additional connections yet, but we are
getting closer!

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vm-migration: allow partitioning memory tables

For sending memory over multiple connections, we need a way to split
up the work. With these changes, we can take a memory table and chop
it into same-sized chunks for transmit.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: actually send memory using multiple connections

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: tear down accept thread after receiving

This will stop us from listening for more connections on the TCP socket
when migration has finished.

Tearing down the individual connections will come in a subsequent
commit.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: tear down memory receive threads

... after the VM migration finishes.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

phip1611 commented on src/bin/ch-remote.rs:
```diff
@@ -489,6 +489,11 @@ fn rest_api_do_command(matches: &ArgMatches, socket: &mut UnixStream) -> ApiResu
+                *matches
+                    .subcommand_matches("send-migration")
+                    .unwrap()
+                    .get_one::<u32>("connections")
```
This could be `NonZeroU32`

blitz replied:
Fixed!

phip1611 commented on src/bin/ch-remote.rs:
```diff
@@ -1102,6 +1114,14 @@ fn get_cli_commands_sorted() -> Box<[Command]> {
+                    .long("connections")
+                    .help("The number of connections to use for the migration")
+                    .num_args(1)
+                    .value_parser(clap::value_parser!(u32))
```
This could be `NonZeroU32`

phip1611 commented on vmm/src/api/mod.rs:
```diff
@@ -258,7 +258,10 @@ pub struct VmReceiveMigrationData {
+    /// URL to migrate the VM to.
+    ///
+    /// This is not actually a URL, but we are stuck with the name. The destination
+    /// is a string, such as tcp:<host>:<port> or unix:/path/to/socket.
```
We aren't stuck with the the name and can change it / break it

blitz replied:
It's part of the HTTP API. I can clarify the comment.

phip1611 commented on vmm/src/api/mod.rs:
```diff
@@ -269,13 +272,21 @@ pub struct VmSendMigrationData {
     #[serde(default)]
     pub migration_timeout: u64,
+    /// The number of parallel connections for migration
+    #[serde(default = "default_connections")]
```
just use `NonZeroU32` and we are good to go I think

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -727,6 +727,39 @@ pub struct Vmm {
 }
 
+/// The receiver's state machine behind the migration protocol.
+enum ReceiveMigrationState {
```
nit: missing Debug impl

blitz replied:
Indeed. This is a rabbit hole to add Debug impls to everything references from `MemoryManager`. :crying_cat_face:

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -727,6 +727,39 @@ pub struct Vmm {
 
+/// The receiver's state machine behind the migration protocol.
+enum ReceiveMigrationState {
+    /// The connection is established and we haven't received any commands yet.
```
`... established and feature negotiation is done ... right?

blitz replied:
There is no feature negotiation anymore. If we still add it, it will be done in the `Started` state. It's unintuitive, but not doing feature negotiation is the key to having this feature work between old and new versions transparently.

phip1611 replied:
k

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -879,6 +912,137 @@ impl Vmm {
     }
 
+    /// Try to receive a file descriptor from a socket. Returns the slot number and the file descriptor.
+    fn vm_receive_memory_fd(
```
Would it be possible to split this change into two commits?

1. extract inline functionality into functions
2. introduce the state machine

blitz replied:
Done. I removed some more `Option<>` while I was at it.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -879,6 +912,137 @@ impl Vmm {
+            return Ok(Aborted);
+        }
+
+        match state {
```
lovely, thanks! Here are some cookies 🍪🍪

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -971,7 +971,7 @@ impl Vmm {
                         for net_config in vm_config.net.iter_mut().flatten() {
                             // update only if the net dev is backed by FDs
-                            if net_config.id == Some(net.id.clone()) && net_config.fds.is_some() {
+                            if net_config.id.as_ref() == Some(&net.id) && net_config.fds.is_some() {
```
I like this, as well as the typo. Could you please extract all unrelated/misc improvements into a dedicated PR?

blitz replied:
This is the only one-liner left. Can you let this slide? :) It just creates merge conflicts to put this into a separate PR and extract it here.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -762,6 +762,125 @@ impl ReceiveListener {
         }
     }
+
+    fn try_clone(&self) -> std::result::Result<Self, std::io::Error> {
```
not about this line but about commit `vmm: handle receiving migration data on additional connections`

`We keep the listening socket around and accept as many questions as
the sender wants to open.` -> probably you mean connections instead of questions?

blitz replied:
Indeed!

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -762,6 +762,125 @@ impl ReceiveListener {
+    }
+}
+
+fn vm_receive_memory<T>(
```
while on it, please add a short description what this function does

blitz replied:
Done!

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -924,6 +924,56 @@ impl ReceiveMigrationState {
+        .map(|s| s.into())
+}
+
+fn send_migration_socket(
```
while on it, please add a short comment what this function is doing

blitz replied:
Done!

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -924,6 +924,56 @@ impl ReceiveMigrationState {
+    }
+}
+
+fn receive_migration_listener(
```
while on it, please add a short comment what this function is doing

blitz replied:
Fixed.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -924,6 +924,101 @@ impl ReceiveMigrationState {
+impl Drop for SendAdditionalConnections {
+    fn drop(&mut self) {
+        // Drop all channels, so the threads will exit.
+        self.channels = vec![];
```
nit. self.channels.clear()

blitz replied:
Fixed.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -924,6 +924,101 @@ impl ReceiveMigrationState {
+        // Drop all channels, so the threads will exit.
+        self.channels = vec![];
+
+        std::mem::take(&mut self.threads)
```
this doesn't look very idiomatic. I think using `.drain(..)` on the vec would be better

blitz replied:
Good point! Fixed.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -965,13 +998,35 @@ impl SendAdditionalConnections {
+                    }
                 }
+                info!(
+                    "Joined thread to send VM memory. Sent {} MiB.",
```
nit: strictly speaking, the thread is done here, but the thread cannot join itself (as the comment indicates)

blitz replied:
I've reworded this a little.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -985,37 +1040,84 @@ impl SendAdditionalConnections {
     /// destination.
+    ///
+    /// When this function returns, all memory has been sent and acknowledged.
     fn send_memory(
```
this handles all memory of a certain memory iteration, right?

blitz replied:
That's how it is currently used. But it doesn't have to be.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -891,6 +900,53 @@ impl ReceiveAdditionalConnections {
 
+    /// Handle read requests. This function returns until the abort_event_fd is
+    /// triggered or the connection is closed or encountered an error.
+    fn abortable_read_requests(
```
from the name of the function, I don't really understand what it is doing

blitz replied:
It's now `handle_requests`.

phip1611 reviewed (changes requested):
Wow, excellent work, Julian! Given the time you had and the state of the code, I think you've came up with an excellent trade-off in refactoring and adding the new functionality. I absolutely love the state machine you introduced! Further, kudos for the awesome git commit hygiene!

Regarding my review: I mostly left some nits. If the open discussion points are resolved, I'm happy to merge this,

phip1611 commented:
Could you split this in three PRs, please?

- unrelated misc improvements
- state machine
- multiple parallel connections?

I think the first two would be merged very easily. I only see potential for further discussions in the last one.

blitz committed:
vmm: gracefully handle remote end not supporting multiple connections

We don't die if we don't manage to establish more than the initial
connection to the other side. To limit the weird failure cases, we do
die when the other side only accepts some connections, but not all.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz committed:
vmm: don't send memory chunked if we only have one connection

This restores the old behavior/performance in case, we don't use
multiple connections.

On-behalf-of: SAP julian.stecklina@sap.com
Signed-off-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz commented:
@phip1611 I've updated the todo list above with the current status. Let's chat tomorrow whether you need anything else for handover.

phip1611 commented:
Regarding the open TODOs by Julian: I think we can ignore some, for example the OpenAPI stuff. It is sufficient if we do this when we upstream this. Regarding the remaining testing: I try to take care of it as part of our hackathon this week

amphi commented on vmm/src/lib.rs:
```diff
@@ -760,6 +1062,304 @@ impl ReceiveMigrationState {
+                otherwise => otherwise,
+            })?;
+            let guest_mem = guest_mem.clone();
+            let (send, recv) = std::sync::mpsc::sync_channel::<SendMemoryThreadMessage>(
+                Self::BUFFERED_REQUESTS_PER_THREAD,
+            );
```
Maybe a stupid question, but: Do I understand it correctly, that you create a mpsc channel per sending thread? And if yes: wouldn't it be more efficient to have some sort of a single producer multiple consumer channel? (like [async channel](https://docs.rs/async-channel/latest/async_channel/)). Then you could just put all partitions into that channel and the threads will take them when they are free.

blitz replied:
That's also possible. I didn't know whether we want to add new dependencies. But this would also solve the problem of buffering too many chunks per thread, which is a problem if one connection becomes slow.

amphi replied:
Yeah I think it would have some advantages. @phip1611 what do you think? Maybe it would be worthwhile to change this?

phip1611 replied:
I think this is a follow-up / next step 🤔

tpressure replied:
i think so too. Please open a ticket

phip1611 replied:
@amphi could you please open a ticket [here](https://github.com/cobaltcore-dev/cobaltcore/issues?q=is%3Aissue%20state%3Aopen%20label%3Acyberus%2Ccyberus-maybe)?

amphi replied:
I opened an issue: https://github.com/cobaltcore-dev/cobaltcore/issues/322

blitz committed:
fixup! vm-migration: add connections field to HTTP API

tpressure commented:
While testing this branch, I realized that we see a

```Error:vmm/src/lib.rs:1006 -- Failed to wake up other threads: Invalid argument (es error 22)```

This happens on every live migration :(

phip1611 commented:
I cannot edit or push to the branch of this PR. Before we waste time fixing this, I open a new one.
